### PR TITLE
Japatel snow 552576 avoid extra memory buffer

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
- Transform record only when records are ready to invoke insertRows API.
- Move around code to only buffer kafka record.
- Add utility method to find out size of kafka record.

PS:
- Suggestions are welcomed for finding out size of kafka record. I couldnt find the best possible way hence this hack. 
- The size of this record is a bit more than when we were just calculating the entire row size. (In Snowpipe Ingestion)
- But this size is still lesser than both row + kafka record combined. 
  - Please note, we have to store Kafka record to be able to send it to kafka DLQ. 

I did the testing:
- Approximate number of records buffered in snowpipe and streaming snowpipe case is same even though the objects buffered in memory are different. 

Most recent commit:

- Process the record and find out the size of serialized record.
- Buffer the original kafka record and throw away the serialized version of kafka which we calculated above.
- Once the threshold has reached, we would go over through the buffered kafka records and serialize them again so as to call insertRows API from streaming snowpipe.
- At the moment, we are not allowed to use open jdk library and we might regress in future if we find accurate object size instead of above mentioned hacky way. (Note that size of object from jol library >> size of serialized record, at least 2-3x)
    - Regress because we might buffer less number of records than before even though size of record might have remained same.
    - Long term solution is to not use buffer bytes at all.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:881]: could we do this in a parallel stream to improve throughput?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java:45]: This is not for record metadata. Record metadata is created after we transform the data. (Done in getData())
Check this in kafka code: https://github.com/apache/kafka/blob/f089bea7ed9fee2bad7084e6b40561ff6935ce5f/clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java#L68

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:211]: Good point.. Let me see if I can send list of records here instead of single record.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:227]: Umm, the log would help us to know if it was size or count based by looking at getBufferSizeBytes or getSinkRecords

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:881]: Yeah sounds like we can use multiple threads here. I dont like parallel streams because it hides away which thread did it and is kind of a blackbox. 
https://stackoverflow.com/questions/20375176/should-i-always-use-a-parallel-stream-when-possible

Although this does qualify for parallelism. 

Would it be okay to send out of order offsets to insertRows? as far as last offset stays the same?

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:881]: I confirmed the code inside this is thread safe to begin with. Also, parallelstream uses 1 less as number of threads from forkjoin pool. So let me just use it..

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:227]: unless you know the threshold, right?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java:45]: ah, thanks, could you add this to the comment to make it more clear

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java:45]: Will do!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:227]: Ahh true. Okay let me see modify.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:881]: No, i think we want to keep the ordering, you will probably need an array and put the processed row to the original index.

File: pom.xml
Comment (sfc-gh-tzhang) [inline pom.xml:441]: Please make sure security is ok with this dependency

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:881]: Let me do it as part of separate PR actually as this is just perf improvement which we cant justify without testing. (In theory yes, it will be better but I want to do testing too) [SNOW-555624](https://snowflakecomputing.atlassian.net/browse/SNOW-555624)

File: pom.xml
Comment (sfc-gh-japatel) [reply pom.xml:441]: I checked the dependency list and they dont have any dependencies except junit. I will still do the security review.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:211]: I feel this is minor improvement which I would want to do as part of [SNOW-555624](https://snowflakecomputing.atlassian.net/browse/SNOW-555624)

Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:211]: sounds good, please add a TODO in the comment

**Review by sfc-gh-tzhang — APPROVED** [review]
lgtm, thanks!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java:120]: This is technically not complete. I should rather do `GraphLayout.parseInstance(kafkaSinkRecord).totalSize()` to get size of entire object..

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingUtils.java:120]: I will change this once we know which library to use.

**Comment by sfc-gh-japatel** [comment]
This is currently blocked on getting an approval to use a library for finding object size.

**Comment by sfc-gh-japatel** [comment]
> This is currently blocked on getting an approval to use a library for finding object size.

I removed the usage of OpenJDK jol library and used an approximation to calculate the running(aggregate) size of buffer. 

IMO, it is a bit hacky at the moment but this is the tradeoff we will have to make. 

Here is what we do in most recent commit:
- Process the record and find out the size of serialized record. 
- Buffer the original kafka record and throw away the serialized version of kafka which we calculated above. 
- Once the threshold has reached, we would go over through the buffered kafka records and serialize them again so as to call insertRows API from streaming snowpipe. 
- At the moment, we are not allowed to use open jdk library and we might regress in future if we find accurate object size instead of above mentioned hacky way. (Note that size of object from jol library >> size of serialized record, at least 2-3x)
  - regress because we might buffer less number of records than before even though size of record might have remained same. 
  - Long term solution is to not use buffer bytes at all. 


Thoughts: @sfc-gh-tzhang @sfc-gh-tjones

**Comment by sfc-gh-tzhang** [comment]
> > This is currently blocked on getting an approval to use a library for finding object size.
> 
> I removed the usage of OpenJDK jol library and used an approximation to calculate the running(aggregate) size of buffer.
> 
> IMO, it is a bit hacky at the moment but this is the tradeoff we will have to make.
> 
> Here is what we do in most recent commit:
> 
> * Process the record and find out the size of serialized record.
> * Buffer the original kafka record and throw away the serialized version of kafka which we calculated above.
> * Once the threshold has reached, we would go over through the buffered kafka records and serialize them again so as to call insertRows API from streaming snowpipe.
> * At the moment, we are not allowed to use open jdk library and we might regress in future if we find accurate object size instead of above mentioned hacky way. (Note that size of object from jol library >> size of serialized record, at least 2-3x)
>   
>   * regress because we might buffer less number of records than before even though size of record might have remained same.
>   * Long term solution is to not use buffer bytes at all.
> 
> Thoughts: @sfc-gh-tzhang @sfc-gh-tjones

Sorry I missed this, but I think this is ok and I won't see this as a regression since Snowpipe Streaming entirely is new, customer could switch back to Snowpipe if they see any issue. In general, as long as we're doing our best to estimate the size then we should be fine. Btw, it should Snowpipe Streaming instead of Streaming Snowpipe :)

**Pull request merged**
